pkg/virt-controller/watch/vm.go

```diff
@@ -924,6 +935,79 @@ func (c *VMController) stopVMI(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMac
 	return nil
 }
 
+func vmRevisionNamePrefix(vmName string) string {
+	return fmt.Sprintf("revision-start-vm-%s", vmName)
+}
+
+func getVMRevisionName(vmName string, generation int64) string {
+	return fmt.Sprintf("%s-%d", vmRevisionNamePrefix(vmName), generation)
+}
+
+func getVMRevisionData(vm *virtv1.VirtualMachine) (runtime.RawExtension, error) {
+	vmCpy := vm.DeepCopy()
+	vmCpy.Status = virtv1.VirtualMachineStatus{}
+	data, err := json.Marshal(vmCpy)
+	if err != nil {
+		return runtime.RawExtension{}, err
+	}
+	return runtime.RawExtension{Raw: data}, nil
+}
+
+func (c *VMController) deleteOlderVMRevision(vm *virtv1.VirtualMachine) (bool, error) {
+	labelSelector := fmt.Sprintf(virtv1.CreatedByLabel + "=" + string(vm.UID))
+	vmRevisions, err := c.clientset.AppsV1().ControllerRevisions(vm.Namespace).List(context.Background(), v1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return false, err
+	}
+
+	createNotNeeded := false
+	for i := range vmRevisions.Items {
+		// check the revision is of the revisions that are created in the vm startup
+		if strings.HasPrefix(vmRevisions.Items[i].Name, vmRevisionNamePrefix(vm.Name)) {
+			if vmRevisions.Items[i].Revision == vm.ObjectMeta.Generation {
+				createNotNeeded = true
+				continue
+			}
+			err = c.clientset.AppsV1().ControllerRevisions(vm.Namespace).Delete(context.Background(), vmRevisions.Items[i].Name, v1.DeleteOptions{})
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+
+	return createNotNeeded, nil
+}
+
+func (c *VMController) createVMRevision(vm *virtv1.VirtualMachine) (string, error) {
+	vmRevisionName := getVMRevisionName(vm.Name, vm.ObjectMeta.Generation)
+	createNotNeeded, err := c.deleteOlderVMRevision(vm)
+	if err != nil || createNotNeeded {
```

Comment:
Please don't mix error checks with other conditions this way if possible as it can result in hard to catch errors.
It's fine here as far as I can see because `vmRevisionName` will never be empty, but in other places (like above) it could result in two empty return values.